### DISCUSSION TITLE
Archway v4.0.0 - smart contract txs fail - incoming patch

### DISCUSSION THREAD

**Discussion by zanicar** [archway-network/discussions#6] [Announcements]
Archway v4.0.0 was released to Constantine today... however, as a result of a missing module capability, transactions are failing for smart contract execution.

We are already working on a patch to address this issue... it will be released shortly so please bear with us.

**Comment by zanicar** [comment]
We are running some final tests and verification internally!
Turns out writing the tests to verify the fix was more complicated than the fix itself...

Archway v4.0.1 (fix) will be tagged in the next couple of hours. Thank you for your patience!

**Reply by shahbazn** [reply]
Fix PR: https://github.com/archway-network/archway/pull/437

**Comment by lufa23** [comment]
Archway v4.0.2 has been released! 

https://github.com/archway-network/archway/releases/tag/v4.0.2

The issue has been addressed, and upgrade proposal coming asap on testnet 🚀

⚠️  Please note that burning the fees consumes gas (12000gas more approximately), meaning that contract executions will be more costly now.  Keep in mind that the more fees burnt, the more value is generated for the protocol.

For further details, please refer to the [Fee Burning ADR](https://github.com/archway-network/archway/blob/main/docs/adr/ADR-007-fee-burning.md)

**Reply by lufa23** [reply]
[Proposal 71](https://testnet.mintscan.io/archway-testnet/proposals/71) is up on Testnet

**Comment by lufa23** [comment]
# Archway v4.0.0 issue and v4.0.2 upgrade proposal
Here some details to better understand what happened

## What has been the issue witnessed with the v4.0.0 archwayd release?

Archway v4.0.0 was released on Constantine on 2023-08-30 around 13:00 UTC. Anyways, because of the missing burn permissions for the fee collector, transactions have been failing for smart contract execution. 
The feeburn feature was introduced in v3.0.0 [#422](https://github.com/archway-network/archway/pull/422), with the objective of making the transaction fees sent with the `wasmTypes.MsgExecuteContract` to be burnt instead of being sent to stakers. More details can be found on the [ADR-007-fee-burning](https://github.com/archway-network/archway/blob/1c062300e134e713a1d210d5cfd98cb92ad332c1/docs/adr/ADR-007-fee-burning.md).
So, while the `authtypes.FeeCollectorName` was given burn permissions in app.go the Fee Collector account was already given the ability to burn tokens in v3.0.0 of archwayd, in the upgrade to v4.0.0 the burn permissions to the Fee Collector account were not properly given, as the upgrade handler was missing the logic to give the permissions.

The discrepancy between the planned state (where the Fee Collector should possess burn permissions) and the realized state (where the Fee Collector lacked burn permissions) let the Fee Collector account without the expected authority to burn tokens, despite the initial intention to enable this capability as part of the upgrade procedure.

This caused the following error on every contract execution, as the burn was not possible:

`Error: Broadcasting transaction failed with code 111222 (codespace: undefined). Log: recovered: module account fee_collector does not have permissions to burn tokens: unauthorized [cosmos/cosmos-sdk@v0.45.16/x/bank/keeper/keeper.go:438]`

## Why a v4.0.2 upgrade proposal was necessary?

As the [proposal 71](https://testnet.mintscan.io/archway-testnet/proposals/71) for the upgrade to the archwayd [v4.0.2](https://github.com/archway-network/archway/releases/tag/v4.0.2) passed, constantine-3 (Archway's testnet) is now functioning correctly.

The version[ v3.0.0](https://github.com/archway-network/archway/releases/tag/v3.0.0) was tagged but never released to testnet/mainnet due to upstream dependency of Cosmwasm receiving an update soon after. Upgrade proposal 71 combines both changes from v3.0.0 and v4.0.2

### What are the upgrade features of v4.0.2?
The upgrade to v4.0.2  adds the following features:

1) [Fee Burn](https://github.com/archway-network/archway/pull/422)
To better accommodate inflationary rewards increases within the Archway protocol, transaction (TX) fee burning is introduced. This strategy permits to offer more inflationary rewards to contracts safely, while simultaneously incinerating the TX fees which were previously distributed to validators and stakers.

Contract Execution txs will consume more gas now (approximately 12000). As such, the fees for Smart Contracts Developers to interact with their contracts will get more expensive. By roughly 0.0108CONST more. It needs to be noted that the more fees burnt, the more value is generated for the protocol.
More details about the fee burn mechanism can be found here: [ADR-007 – Introduction of fee burning](https://github.com/archway-network/archway/blob/0f74dfbf6178578bc30fb2cbee3ae1ed5bc63dd9/docs/adr/ADR-007-fee-burning.md)

2) [Cosmwasm_1_3 capabilities](https://github.com/archway-network/archway/pull/429)
By bumping the wasmd to v0.33.0, archway now supports [cosmwasm_1_3 capabilities](https://github.com/CosmWasm/cosmwasm/blob/f3d57ce2a92c55495267c426e3c9c0364bf98596/docs/CAPABILITIES-BUILT-IN.md).
See the [Full Change Log](https://github.com/archway-network/archway/blob/v4.0.2/CHANGELOG.md) here.

**Comment by 7Two1** [comment]
> constantine-3 (Archway's testnet) is now functioning correctly
